### PR TITLE
fix: calculate db and collection tags only once for streaming requests

### PR DESCRIPTION
**Pull request by pboros**
Calculate db and collection tagging only once.

### CODE REVIEW

File: server/middleware/measure.go [hunk 1]
```diff
@@ -117,37 +117,51 @@ func measureStream() grpc.StreamServerInterceptor {
...
-	reqMetadata, err1 := request.GetRequestMetadataFromContext(w.WrappedContext)
-	if err1 != nil {
-		return errors.Internal("Could not handle stream send message")
+
+	if len(w.measurement.GetDBCollTags()) == 0 {
+		// The request is not tagged yet with db and collection, need to do it on the first message
+		db, coll := request.GetDbAndColl(m)
+		reqMetadata, err := request.GetRequestMetadataFromContext(w.WrappedContext)
+		if err != nil {
+			ulog.E(err)
```
Comment (rbarabas) [inline server/middleware/measure.go:158]: Could this potentially blow up our logging if it gets triggered on `SendMsg()`?

Comment (rbarabas) [reply server/middleware/measure.go:158]: Perhaps we could set the error flag on measurement to avoid logging N times?

Comment (pboros) [reply server/middleware/measure.go:158]: Yes, this is for each document in the result set. One may have error, the other may not. We can also just log it at debug level or something like that.

File: server/middleware/measure.go [hunk 2]
```diff
@@ -117,37 +117,51 @@ func measureStream() grpc.StreamServerInterceptor {
...
-	db, coll := request.GetDbAndColl(m)
-	reqMetadata, err := request.GetRequestMetadataFromContext(w.WrappedContext)
-	if err != nil {
-		return err
+
+	if len(w.measurement.GetDBCollTags()) == 0 {
+		// The request is not tagged yet with db and collection, need to do it on the first message
+		db, coll := request.GetDbAndColl(m)
+		reqMetadata, err := request.GetRequestMetadataFromContext(w.WrappedContext)
+		if err != nil {
```
Comment (adilansari) [inline server/middleware/measure.go:132]:
```suggestion
		if ulog.E(err) {
				return err 
		}
```

File: server/middleware/measure.go [hunk 3]
```diff
@@ -117,37 +117,51 @@ func measureStream() grpc.StreamServerInterceptor {
...
+			ulog.E(err)
```
Comment (himank) [inline server/middleware/measure.go:133]: if err is not nil can we still safely tag it in line `135`?

File: server/middleware/measure.go [hunk 4]
```diff
@@ -117,37 +117,51 @@ func measureStream() grpc.StreamServerInterceptor {
 }
 
 func (w *wrappedStream) RecvMsg(m interface{}) error {
+	err := w.ServerStream.RecvMsg(m)
+	if err != nil {
+		ulog.E(err)
```
Comment (himank) [inline server/middleware/measure.go:122]: why not just return it from here?

File: server/middleware/measure.go [hunk 5]
```diff
@@ -117,37 +117,51 @@ func measureStream() grpc.StreamServerInterceptor {
...
-		return errors.Internal("Could not handle stream send message")
+
+	if len(w.measurement.GetDBCollTags()) == 0 {
+		// The request is not tagged yet with db and collection, need to do it on the first message
+		db, coll := request.GetDbAndColl(m)
+		reqMetadata, err := request.GetRequestMetadataFromContext(w.WrappedContext)
+		if err != nil {
+			ulog.E(err)
+		}
+		reqMetadata.SetDb(db)
```
Comment (himank) [inline server/middleware/measure.go:160]: same as above, if there is an error above is it safe to use `reqMetadata` object?

File: server/middleware/measure.go (hunk 3 above)
Comment (pboros) [reply server/middleware/measure.go:133]: Right, we can't, we need to return the error here and maybe have a log message that this request will not be measured.

File: server/middleware/measure.go (hunk 2 above)
Comment (pboros) [reply server/middleware/measure.go:132]: Thanks, good catch, but here we need to return the error or the original recv call, not the one from getting request metadata. Changing the code so it's more readable.

File: server/middleware/measure.go (hunk 4 above)
Comment (pboros) [reply server/middleware/measure.go:122]: I had trouble with that, it breaks things. I wanted to do it like the SendMsg (it returns error there), but it even breaks the tests.

File: server/middleware/measure.go (hunk 5 above)
Comment (pboros) [reply server/middleware/measure.go:160]: Right it's not, we should return the original error here.

**Review by himank — DISMISSED** [review]

**Review by himank — APPROVED** [review]

**Pull request merged**